### DISCUSSION TITLE
serialization of vectors and user_data

### DISCUSSION THREAD

**Discussion by the1gofer** [explosion/spaCy#8719] [Help: Coding & Implementations] [answered]
Is it possible to serialize the doc object with vectors and user_data?    

It seems like to_bytes leaves out vectors,  and pickle leaves out user_data.     Maybe I'm misunderstanding.  

I've got a bunch of documents I need to analyse, and would like to save the doc object so I don't have to spend the time recreating it every time I come back to it later.  I read that you recommend doing DocBin for large collections, but I can't do that if I have no way to reference it back my original data somehow.

**Comment by adrianeboyd** [comment, marked as answer]
The vectors are part of the vocab and not the doc annotation itself, and usually you wouldn't want to save them with the doc because it would be a lot of repeated data. If you use `pickle`, a copy of the entire vector table is saved with each doc and each doc will be really really huge. In general, we'd almost never recommend using `pickle` for this.

We would recommend using `DocBin` with `store_user_data=True` (you can easily add an additional custom extension that links each doc to an ID or metadata if that's an issue) and then when you load the docs back, use the same pipeline for the provided vocab so you have access to the same vectors:

```python
nlp = spacy.load("en_core_web_lg")
docs = DocBin().from_disk("file.spacy").get_docs(nlp.vocab)
for doc in docs:
    ...
```

If you really want the vectors for a particular doc to be saved with it in an independent way, then you could consider copying the vectors into a custom extension, but you'd have to deal with copying the vectors back into the vocab after reloading to be able to use the similarity methods (without further changes anyway, I could imagine alternatives with custom vector hooks). And if you copy vectors back, you'd have to be careful if you're mixing docs from different sources because the vocab can only manage one vector per token form.

**Reply by the1gofer** [reply]
Thanks.  I'll give that a shot.
